core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -504,17 +500,34 @@ public void run() {
 
   @Override
   public final void cancel(final Status reason) {
-    Substream noopSubstream = new Substream(0 /* previousAttempts doesn't matter here */);
-    noopSubstream.stream = new NoopClientStream();
-    Runnable runnable = commit(noopSubstream);
+    if (isHedging) {
+      Substream noopSubstream = new Substream(0 /* previousAttempts doesn't matter here */);
+      noopSubstream.stream = new NoopClientStream();
+      Runnable runnable = commit(noopSubstream);
 
-    if (runnable != null) {
+      if (runnable != null) {
+        synchronized (lock) {
+          state = state.substreamDrained(noopSubstream);
+        }
+        runnable.run();
+        safeCloseMasterListener(reason, RpcProgress.PROCESSED, new Metadata());
+        return;
+      }
+    } else {
+      Optional<Substream> substreamToCommit;
```

Comment:
We can't use optional. We support Android versions that don't have it. Ditto to stream(). We'd use `Iterables.getFirst(state.drainedSubstreams, null)`. Animalsniffer should catch these things.